# If-Statements with User-Defined Functions

In [1]:
# importing libraries
import pandas as pd
import numpy as np
import os

In [5]:
# importing dataframe
path = r'/Users/sudip/Desktop/DATA ANALYST/Projects/Project 4 Python/Instacart Basket Analysis'
df_ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [8]:
# creating subset of the dataframe for first one million rows.
df = df_ords_prods_merged[:1000000]

In [9]:
df.shape

(1000000, 16)

In [14]:
# define a function
def price_label(row):
    if row['prices'] <= 5:
        return 'Low range product'
    if (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid range product'
    if row['prices'] > 15:
        return 'High range'
    else: return 'Not enough Data'

In [15]:
#Apply the function
df['price_label'] = df.apply(price_label, axis = 1)

<ipython-input-15-597830539bb0>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_label'] = df.apply(price_label, axis = 1)


In [17]:
df.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   
1                    15.0         False         196                  1   
2                    21.0         False         196                  1   
3                    29.0         False         196                  1   
4                    28.0         False         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

         price_label  
0  Mid range product  
1  Mid range product  
2  Mid range product  
3  Mid range product  
4  Mid range product

In [18]:
df['price_label'].value_counts(dropna = False)

Mid range product    756450
Low range product    243550
Name: price_label, dtype: int64

In [19]:
df['prices'].max()

14.8

In [20]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices']> 15, 'price_range_loc'] = 'High_range product'

In [22]:
df_ords_prods_merged.loc[(df_ords_prods_merged['prices']<=15) & (df_ords_prods_merged['prices']> 5), 'price_range_loc'] = 'Mid_range product'

In [24]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] <= 5, 'price_range_loc'] = 'Low_range product'

In [25]:
df_ords_prods_merged['price_range_loc'].value_counts(dropna = False)

Mid_range product     21860860
Low_range product     10126321
High_range product      417678
Name: price_range_loc, dtype: int64

In [28]:
df_ords_prods_merged = df_ords_prods_merged.rename(columns = {'order_dow': 'orders_day_of_the_week'})

In [29]:
df_ords_prods_merged['orders_day_of_the_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_the_week, dtype: int64

# If statements with for-loops

In [30]:
#creating for-loops
result = []
for value in df_ords_prods_merged['orders_day_of_the_week']:
    if value == 0:
        result.append('Busiest Day')
    elif value == 4:
        result.append('Least Busy')
    else :
        result.append('Regularly Busy')

In [31]:
result

['Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Least Busy',
 'Least Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Least Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Least Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Least Busy',
 'Least Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Regularly Busy',
 'Least Busy',
 'Regularly Busy',
 'Busiest Day',
 'Regularly Busy',
 'Reg

In [32]:
df_ords_prods_merged['busiest_day'] = result

In [33]:
df_ords_prods_merged['busiest_day'].value_counts(dropna = False)

Regularly Busy    22416875
Busiest Day        6204182
Least Busy         3783802
Name: busiest_day, dtype: int64

# Task 4.7

In [37]:
# creating for-loop for busiest days and slowest days
result2 = []

for value in df_ords_prods_merged['orders_day_of_the_week']:
  if value in [0,1]:
    result2.append("Busiest days")
  elif value in [3,4]:
    result2.append("Slowest days")
  else:
    result2.append("Regularly busy")

In [38]:
df_ords_prods_merged['busiest_days'] = result2

In [39]:
df_ords_prods_merged['busiest_days'].value_counts(dropna= False)

Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: busiest_days, dtype: int64

The total number of rows for 0 and 1 are 6204182 and 5660230, which is equal to busiest days rows 11864412. The total number of rows for 3 and 4 are 3840534 and 3783802, which is equal to Slowest days rows 7624336. So the values in this new columns are accurate.

In [42]:
#dentifying the busiest hours of the day
df_ords_prods_merged['order_hour_of_day'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: order_hour_of_day, dtype: int64

In [43]:
# Creating for-loop for sorting the busiest hours of the day by periods
result3 = []
for value in df_ords_prods_merged['order_hour_of_day']:
    if value >=9 and value <=16:
        result3.append('Most orders')
    elif value >=0 and value <=5:
        result3.append('Fewest orders')
    else:
        result3.append('Average orders')

In [44]:
# Creating new column 'busiest_period_of_day' with the results of the for-loop
df_ords_prods_merged['busiest_period_of_day'] = result3

In [45]:
df_ords_prods_merged['busiest_period_of_day'].value_counts(dropna = False)

Most orders       21118071
Average orders    10690460
Fewest orders       596328
Name: busiest_period_of_day, dtype: int64

In [46]:
df_ords_prods_merged.head()

order_id  user_id eval_set  order_number  orders_day_of_the_week  \
0   2539329        1    prior             1                       2   
1   2398795        1    prior             2                       3   
2    473747        1    prior             3                       3   
3   2254736        1    prior             4                       4   
4    431534        1    prior             5                       4   

   order_hour_of_day  days_since_prior_order  new_customer  product_id  \
0                  8                     NaN          True         196   
1                  7                    15.0         False         196   
2                 12                    21.0         False         196   
3                  7                    29.0         False         196   
4                 15                    28.0         False         196   

   add_to_cart_order  reordered product_name  aisle_id  department_id  prices  \
0                  1          0         Soda        77              7     9.0   
1                  1          1         Soda        77              7     9.0   
2                  1          1         Soda        77              7     9.0   
3                  1          1         Soda        77              7     9.0   
4                  1          1         Soda        77              7     9.0   

  _merge    price_range_loc     busiest_day    busiest_days  \
0   both  Mid_range product  Regularly Busy  Regularly busy   
1   both  Mid_range product  Regularly Busy    Slowest days   
2   both  Mid_range product  Regularly Busy    Slowest days   
3   both  Mid_range product      Least Busy    Slowest days   
4   both  Mid_range product      Least Busy    Slowest days   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2           Most orders  
3        Average orders  
4           Most orders

In [47]:
#exporting the dataframe in pickle file
df_ords_prods_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_derived.pkl'))